### PR TITLE
bugfix: throw error if T_END found in the middle of input

### PR DESCRIPTION
**Pull request by fffonion**
KAG-4275

### CODE REVIEW

File: lua_cjson.c [hunk 1]
```diff
@@ -1441,6 +1441,10 @@ static int json_decode(lua_State *l)
     if (token.type != T_END)
         json_throw_parse_error(l, &json, "the end", &token);
 
+    /* Make sure T_END (\x00) doesn't occur at middle of input */
```
Comment (chronolaw) [inline lua_cjson.c]:
```suggestion
    /* Make sure T_END (\x00) doesn't occur at the middle of input */
```

**Comment by chronolaw** [comment]
Should we back port this PR to upstream cjson？

**Comment by fffonion** [comment]
@chronolaw yes we should, will do after 3.7 release

**Review by chronolaw — APPROVED** [review]

File: lua_cjson.c [hunk 2]
```diff
@@ -1441,6 +1441,10 @@ static int json_decode(lua_State *l)
     if (token.type != T_END)
         json_throw_parse_error(l, &json, "the end", &token);
 
+    /* Make sure T_END (\x00) doesn't occur at the middle of input */
+    if (json.data + json_len > json.ptr)
+        json_throw_parse_error(l, &json, "EOF", &token);
```
Comment (vm-001) [inline lua_cjson.c:1446]: Is making this error handling become an instance configuration a good idea? 

Most JSON libraries from various languages can fail with an error in this case, and I believe this is a bug. However, it's sitting there for a very long time(in Kong Gateway)... : ( 

This PR is going to fix it entirely like a nuclear bomb. All `cjson.decode` in Kong Gateway will be affected, even the code from custom plugins and custom patches.

**Review by chobits — APPROVED** [review]

**Pull request merged**